### PR TITLE
No alloc FileEntity from workflow_get_update_file

### PR DESCRIPTION
**Pull request by jw-msft**
workflow_get_update_file and workflow_get_update_file_by_name unnecessarily allocate the FileEntity for the out param. Fixed to just pass in address of stack variable instead to avoid fragmenting the heap.

### CODE REVIEW

**Review by yuvalg-MSFT — DISMISSED** [review]

File: src/utils/workflow_utils/inc/aduc/workflow_utils.h [hunk 1]
```diff
@@ -233,20 +233,20 @@ size_t workflow_get_update_files_count(ADUC_WorkflowHandle handle);
  *
  * @param handle A workflow data object handle.
  * @param index An index of the file to get.
- * @param entity An output file entity object. Caller must free the object with workflow_free_file_entity().
- * @return true If succeeded.
+ * @param entity An output file entity object. Caller must uninitialize it via ADUC_FileEntity_Uninit when done.
```
Comment (JeffMill) [inline src/utils/workflow_utils/inc/aduc/workflow_utils.h:236]: This PR is a great start, but if you notice, the bug title I created is "workflow_get*" -- there's a lot of other methods in here that are returning ADUC_FileEntity**.

any chance of getting those changed as well?

**Review by JeffMill — CHANGES_REQUESTED** [review]

File: src/utils/workflow_utils/inc/aduc/workflow_utils.h (hunk 1 above)
Comment (jw-msft) [reply src/utils/workflow_utils/inc/aduc/workflow_utils.h:236]: I had missed that. In that case, this is a larger refactoring task that I can continue on post .2 patch.

Comment (JeffMill) [reply src/utils/workflow_utils/inc/aduc/workflow_utils.h:236]: can you create a new bug/task for that?

File: src/extensions/step_handlers/script_handler/src/script_handler.cpp [hunk 2]
```diff
@@ -225,17 +226,15 @@ ADUC_Result ScriptHandlerImpl::Download(const tagADUC_WorkflowData* workflowData
                 .retryTimeout = DO_RETRY_TIMEOUT_DEFAULT,
             };
 
-            result = ExtensionManager::Download(entity, workflowHandle, &downloadOptions, nullptr);
+            result = ExtensionManager::Download(&fileEntity, workflowHandle, &downloadOptions, nullptr);
+            ADUC_FileEntity_Uninit(&fileEntity);
         }
```
Comment (Nox-MSFT) [inline src/extensions/step_handlers/script_handler/src/script_handler.cpp:231]: Do we need this? We already call ADUC_FileEntity_Uninit at the end of the function.

**Review by Nox-MSFT — APPROVED** [review]

File: src/utils/workflow_utils/inc/aduc/workflow_utils.h (hunk 1 above)
Comment (jw-msft) [reply src/utils/workflow_utils/inc/aduc/workflow_utils.h:236]: I went ahead and fixed remaining--wasn't too bad.

File: src/extensions/step_handlers/script_handler/src/script_handler.cpp (hunk 2 above)
Comment (jw-msft) [reply src/extensions/step_handlers/script_handler/src/script_handler.cpp:231]: removed

File: src/adu_workflow/src/agent_workflow.c [hunk 3]
```diff
@@ -1169,14 +1168,21 @@ static void CallDownloadHandlerOnUpdateWorkflowCompleted(const ADUC_WorkflowHand
     for (size_t i = 0; i < payloadCount; ++i)
     {
         ADUC_Result result = {};
-        ADUC_FileEntity* fileEntity = NULL;
-        if (!workflow_get_update_file(workflowHandle, i, &fileEntity) || IsNullOrEmpty(fileEntity->DownloadHandlerId))
+        ADUC_FileEntity fileEntity;
+        memset(&fileEntity, 0, sizeof(fileEntity));
```
Comment (JeffMill) [inline src/adu_workflow/src/agent_workflow.c:1172]: nit: zero not required.  

`foo f;
if (get_f(&f()) { ... }`

doesn't access f on failure, and should assume that get_f has fully intited f.

Comment (JeffMill) [reply src/adu_workflow/src/agent_workflow.c:1172]: (throughout)

File: src/extensions/download_handlers/plugin_examples/microsoft_delta_download_handler/source_update_cache/src/source_update_cache_utils.c [hunk 4]
```diff
@@ -149,7 +150,8 @@ ADUC_Result ADUC_SourceUpdateCacheUtils_MoveToUpdateCache(
 {
     ADUC_Result result = { .ResultCode = ADUC_Result_Failure };
     int res = -1;
-    ADUC_FileEntity* fileEntity = NULL;
+    ADUC_FileEntity fileEntity;
+    memset(&fileEntity, 0, sizeof(fileEntity));
```
Comment (JeffMill) [inline src/extensions/download_handlers/plugin_examples/microsoft_delta_download_handler/source_update_cache/src/source_update_cache_utils.c:154]: note that in cases like this, you DO want to init fileEntity, as you're unconditionally calling ADUC_FileEntity_Uninit in done, so maybe it's best to always zero it out.

File: src/adu_workflow/src/agent_workflow.c (hunk 3 above)
Comment (JeffMill) [reply src/adu_workflow/src/agent_workflow.c:1172]: see my note below about when we *do* need the zero init.

Also, consider (not essential) changing the instances of memset() to:

    ADUC_FileEntity_Init(&fileEntity);

which just does a zero init currently.  It matches the Uninit call.

**Review by JeffMill — APPROVED** [review]
approved with comment.

**Review by Nox-MSFT — APPROVED** [review]

**Review by Nox-MSFT — APPROVED** [review]

File: src/adu_workflow/src/agent_workflow.c (hunk 3 above)
Comment (jw-msft) [reply src/adu_workflow/src/agent_workflow.c:1172]: workflow_get_update_file already calls ADUC_FileEntity_init on caller's behalf (line 1808), setting out param as a side-effect.

File: src/extensions/download_handlers/plugin_examples/microsoft_delta_download_handler/source_update_cache/src/source_update_cache_utils.c (hunk 4 above)
Comment (jw-msft) [reply src/extensions/download_handlers/plugin_examples/microsoft_delta_download_handler/source_update_cache/src/source_update_cache_utils.c:154]: agreed. as long as it's zero'ed out then the uninit will noop.

**Pull request merged**